### PR TITLE
Fix existing ns issue

### PR DESCRIPTION
**Pull request by lexfrei**
## Description
- Fix for #1740 
- Respect namespace helm variable

## Todos
- [ ] Tests
- [x] Documentation
- [x] Release note

## Release Note

```release-note
None required
```

### CODE REVIEW

**Comment by rbrtbnfgl** [comment]
Are you sure that this will fix the issue? It seems relate to the namespace with this fix using helm to create the namespace will lead to an issue when helm is trying to create the namespace resource.
On your setup when you used `helmfile` is not enough to not specify the namespace.

**Comment by lexfrei** [comment]
Yep, it looks like some magic happened with `helmfile` in the previous iteration. Let's leave it and discuss just the bare `helm` :)

For now, the flannel's documentation is not working for me:

```bash
a.sviridkin@mbp:~/git/github.com/lexfrei/k8s$ kubectl delete ns kube-flannel
Error from server (NotFound): namespaces "kube-flannel" not found
✗: 1 @ Wed Mar 15 15:33:26 MSK 2023
a.sviridkin@mbp:~/git/github.com/lexfrei/k8s$ helm install flannel --set podCidr="10.244.0.0/16" https://github.com/flannel-io/flannel/releases/latest/download/flannel.tgz --set flannel.backend="wireguard"
Error: INSTALLATION FAILED: create: failed to create: namespaces "kube-flannel" not found
✗: 1 @ Wed Mar 15 15:33:33 MSK 2023
```

However, we can't use native helm's `--create namespace`, coz templated namespace name will conflict with the existing one:

```bash
a.sviridkin@mbp:~/git/github.com/lexfrei/k8s$ helm upgrade --install flannel --namespace kube-flannel --set podCidr="10.244.0.0/16" --set flannel.backend="wireguard" https://github.com/flannel-io/flannel/releases/latest/download/flannel.tgz
Release "flannel" does not exist. Installing it now.
Error: namespaces "kube-flannel" already exists
✗: 1 @ Wed Mar 15 15:37:53 MSK 2023
```

So, we need to create the namespace first:

```bash
a.sviridkin@mbp:~/git/github.com/lexfrei/k8s$ kubectl create ns kube-flannel
namespace/kube-flannel created
✓ @ Wed Mar 15 15:57:43 MSK 2023
a.sviridkin@mbp:~/git/github.com/lexfrei/k8s$ kubectl label ns kube-flannel pod-security.kubernetes.io/enforce=privileged
namespace/kube-flannel labeled
✓ @ Wed Mar 15 15:57:47 MSK 2023
a.sviridkin@mbp:~/git/github.com/lexfrei/k8s$ helm install flannel --namespace kube-flannel --set podCidr="10.244.0.0/16" --set flannel.backend="wireguard" ./flannel
NAME: flannel
LAST DEPLOYED: Wed Mar 15 15:57:59 2023
NAMESPACE: kube-flannel
STATUS: deployed
REVISION: 1
TEST SUITE: None
✓ @ Wed Mar 15 15:58:00 MSK 2023
```

Some research shows that helm project maintainers don't want to support namespace creation with labeling, so we need to do it manually. (closed as "won't fix" here: https://github.com/helm/helm/issues/3503#issuecomment-541335015)
My suggestion is to use `kubectl` for namespace creation and `helm` for the rest. So we need to delete the namespace from the chart and add `kubectl` command to the docs.

Another option is to use magic pre-hook, but I don't like it, because it's not obvious, is against the official helm practices and requires some research to understand what's going on.

P.S. My previous attempt to add helm labels to ns is working for adoption, but not for creation. So, it's not a solution or a simplification.

**Comment by lexfrei** [comment]
Here are example tests for daemonset

File: README.md [hunk 1]
```diff
@@ -34,23 +34,28 @@ Flannel can be added to any existing Kubernetes cluster though it's simplest to
...
 
 If you use custom `podCIDR` (not `10.244.0.0/16`) you first need to download the above manifest and modify the network to match your one.
 
 #### Deploying Flannel with helm
-```
-helm install flannel --set podCidr="10.244.0.0/16" https://github.com/flannel-io/flannel/releases/latest/download/flannel.tgz
+
+```bash
+kubectl create ns kube-flannel
+kubectl label ns kube-flannel pod-security.kubernetes.io/enforce=privileged
```
Comment (rbrtbnfgl) [inline README.md:48]:
```suggestion
kubectl label --overwrite ns kube-flannel pod-security.kubernetes.io/enforce=privileged
```

File: chart/kube-flannel/templates/config.yaml [hunk 2]
```diff
@@ -2,7 +2,6 @@ kind: ConfigMap
 apiVersion: v1
 metadata:
   name: kube-flannel-cfg
-  namespace: kube-flannel
```
Comment (rbrtbnfgl) [inline chart/kube-flannel/templates/config.yaml:5]: I know that using `--namespace` will automatically add the resources on the configured namespace but I prefer to specify `namespace: {{ .Release.Namespace }}`

File: chart/kube-flannel/tests/daemonset_test.yaml [hunk 3]
```diff
@@ -0,0 +1,47 @@
+suite: test daemonset
```
Comment (rbrtbnfgl) [inline chart/kube-flannel/tests/daemonset_test.yaml:1]: How does this test work? Do we have to package it inside the chart or we can skip it with `.helmignore`?

**Review by rbrtbnfgl — CHANGES_REQUESTED** [review]

File: chart/kube-flannel/templates/config.yaml (hunk 2 above)
Comment (lexfrei) [reply chart/kube-flannel/templates/config.yaml:5]: Done

File: README.md (hunk 1 above)
Comment (lexfrei) [reply README.md]: Done

File: chart/kube-flannel/tests/daemonset_test.yaml [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+          value: kube-flannel-ds
+
+  - it: should have the correct image
+    set:
+      flannel.image.repository: cicker.io/flannel/flannel
+      flannel.image.tag: v555.44.3
+    asserts:
+      - equal:
+          path: spec.template.spec.containers[0].image
+          value: cicker.io/flannel/flannel:v555.44.3
```
Comment (rbrtbnfgl) [inline chart/kube-flannel/tests/daemonset_test.yaml:22]: What is this repo?

Comment (lexfrei) [reply chart/kube-flannel/tests/daemonset_test.yaml:22]: The not existing one, just to check the value is passed correctly and it's not passed from values or else. The same is the not existing tag.

File: chart/kube-flannel/tests/daemonset_test.yaml (hunk 3 above)
Comment (lexfrei) [reply chart/kube-flannel/tests/daemonset_test.yaml:1]: Should I do something about it?

Comment (rbrtbnfgl) [reply chart/kube-flannel/tests/daemonset_test.yaml:1]: No it's ok.

File: chart/kube-flannel/.helmignore [hunk 5]
```diff
@@ -0,0 +1 @@
+tests/
```
Comment (rbrtbnfgl) [inline chart/kube-flannel/.helmignore:1]: Could you also add `.helmignore` between the files to ignore when creating the chart.
```suggestion
.helmignore
tests/
```

Comment (lexfrei) [reply chart/kube-flannel/.helmignore:1]: done

**Comment by lexfrei** [comment]
Should I bring tests other components in another PR?

**Comment by rbrtbnfgl** [comment]
No the tests that you have included are fine. I think I will add the tests with the automatic tests.

**Pull request merged**